pkg/controller/operators/catalog/operator.go

```diff
@@ -2013,81 +2013,101 @@ func transitionInstallPlanState(log logrus.FieldLogger, transitioner installPlan
 	logrus.Debugf("Comparing %#v to %#v", oldCRD.Spec.Validation, newCRD.Spec.Validation)
 
-	// TODO return early of all versions are equal
-	convertedCRD := &apiextensions.CustomResourceDefinition{}
-	if err := apiextensionsv1beta1.Convert_v1beta1_CustomResourceDefinition_To_apiextensions_CustomResourceDefinition(newCRD, convertedCRD, nil); err != nil {
-		return err
+	oldVersionSet := make(map[string]struct{})
+	if len(oldCRD.Spec.Versions) == 0 {
+		// apiextensionsv1beta1 special case: if spec.Versions is empty, use the global version and validation
+		oldVersionSet[oldCRD.Spec.Version] = struct{}{}
 	}
-	for _, version := range oldCRD.Spec.Versions {
-		if version.Served {
-			gvr := schema.GroupVersionResource{Group: oldCRD.Spec.Group, Version: version.Name, Resource: oldCRD.Spec.Names.Plural}
-			err := validateExistingCRs(dynamicClient, gvr, convertedCRD)
-			if err != nil {
+	for _, oldVersion := range oldCRD.Spec.Versions {
+		// collect served versions from spec.Versions if the list is present
+		if _, ok := oldVersionSet[oldVersion.Name]; !ok && oldVersion.Served {
+			oldVersionSet[oldVersion.Name] = struct{}{}
+		}
+	}
+
+	validationsMap := make(map[string]*apiextensions.CustomResourceValidation, 0)
+	if len(newCRD.Spec.Versions) == 0 {
+		// apiextensionsv1beta1 special case: if spec.Versions of newCRD is empty, use the global version and validation
+		if _, ok := oldVersionSet[newCRD.Spec.Version]; ok {
+			convertedValidation := &apiextensions.CustomResourceValidation{}
+			if err := apiextensionsv1beta1.Convert_v1beta1_CustomResourceValidation_To_apiextensions_CustomResourceValidation(newCRD.Spec.Validation, convertedValidation, nil); err != nil {
 				return err
 			}
+			validationsMap[newCRD.Spec.Version] = convertedValidation
 		}
 	}
-
-	if oldCRD.Spec.Version != "" {
-		gvr := schema.GroupVersionResource{Group: oldCRD.Spec.Group, Version: oldCRD.Spec.Version, Resource: oldCRD.Spec.Names.Plural}
-		err := validateExistingCRs(dynamicClient, gvr, convertedCRD)
-		if err != nil {
-			return err
+	for _, newVersion := range newCRD.Spec.Versions {
+		if _, ok := oldVersionSet[newVersion.Name]; ok && newVersion.Served {
+			// If the new CRD's version is present in the cluster and still
+			// served then fill the map entry with the new validation
+			schema := newVersion.Schema
+			if schema == nil {
+				// apiextensionsv1beta1 special case: if per-version schema is absent, use the global validation
+				schema = newCRD.Spec.Validation
+			}
+			convertedValidation := &apiextensions.CustomResourceValidation{}
+			if err := apiextensionsv1beta1.Convert_v1beta1_CustomResourceValidation_To_apiextensions_CustomResourceValidation(schema, convertedValidation, nil); err != nil {
+				return err
+			}
+			validationsMap[newVersion.Name] = convertedValidation
```

Comment:
You can either have per-version schemas, or a global schema, but not a mix. It might be worth adjusting this a bit to take that into account. In other words, if you had this:

- oldVersion and newVersion both have:
  - 3 versions, no schema for any of them
  - global schema validation

I think the way it's currently coded, it will use the same global validation for all 3 versions, making 3 list calls, and validating everything 3 times against the same global schema... right?

Reply:
That's right; good point. I've updated the loop to validate once and return early when the global validation is set. This should prevent unnecessary lists and validations.